third_party/move/move-compiler-v2/tests/checking/typing/freeze_mut_ref.move

```diff
@@ -0,0 +1,52 @@
+module 0x42::freeze_mut_ref {
+    use std::vector;
+
+    public fun borrow_mut<Element>(
+        map: &mut vector<Element>,
+    ): &Element {
+        vector::borrow_mut(map, 0)
+    }
+
+    public fun borrow_mut2<Element>(
+        v: &mut Element,
+    ): &Element {
+        v
+    }
+
+    public fun borrow_mut3<Element>(
+        v1: &mut Element,
+        v2: & Element
+    ): &Element {
+        if (true)
+            v1
+        else
+            v2
+    }
+
+    public fun borrow_mut4<Element>(
+        v: &mut Element,
+    ): &Element {
+        return v
+    }
+
+    struct S {
+
+    }
+
+    fun t0() {
+        let x: &u64 = &mut 0; x;
+    }
+
+    fun t1(s: &mut S): &S {
+        s
+    }
+
+    fun t2(u1: &mut u64, u2: &mut u64): (&u64, &mut u64) {
+        (u1, u2)
+    }
+
+    fun t3(s: &mut S) {
```

Comment:
If you add this code to transactional tests, you will notice that `S` needs drop ability. If we add drop ability, then v2 successfully processes this code, but v1 throws a compiler error, expecting s `to` be a mut ref.

This is a behavioral change, and we may want to explicitly call out whether we want to behave differently than v1 here (and have tests etc to demonstrate that).

Reply:
Yeah, this is actually a bug revealed by another test case `mutable_immutable.mov`. I will file an issue for it.

Reply:
A fix is added to this PR